internal/adapters/currencyrate/readers/currency_api.go

```diff
@@ -0,0 +1,39 @@
+package readers
+
+import (
+	"fmt"
+	"log"
+	"strings"
+)
+
+type CurrencyAPICurrencyReader struct {
+	APIURL string
+}
+
+func NewCurrencyAPICurrencyReader(url string) *CurrencyAPICurrencyReader {
+	if url == "" {
+		log.Printf("WARNING: CurrencyAPICurrencyReader: url is empty")
+		return nil
+	}
+	return &CurrencyAPICurrencyReader{
+		APIURL: url,
+	}
+}
+
+func (cr *CurrencyAPICurrencyReader) GetCurrencyRate(from string, to string) (float32, error) {
+	data, err := ReadHTTP(cr.APIURL + "&base_currency=" + strings.ToUpper(from))
+	if err != nil {
+		return 0, err
+	}
+
+	if rates, ok := (*data)["data"].(map[string]interface{}); ok {
```

Comment:
Can we use struct here?

Reply:
We can't use structs here, because a response's body is not unified

Reply:
But, than I didn`t understand the logic.

You have three independent providers. All of them have method `GetCurrencyRate` and can encapsulate `getRates` with specific struct in them, didn't you?


<img width="916" alt="image" src="https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-AndrewSergienko/assets/25195930/7252c1ac-5610-4401-b6e2-22d8bd8d1815">